### PR TITLE
Fix issue 17075 ctRegex BacktrackingMatcher.prevStack: free(): invali…

### PR DESCRIPTION
**Pull request by DmitryOlshansky**
Backtracking engine memory management was horribly broken in multiple ways.
This fixes the mentioned issue and a couple other subtle bugs.

Finally I can pass the test suite with memory chunk size of 1 state, so things are more robust.

### CODE REVIEW

File: std/regex/internal/backtracking.d [hunk 1]
```diff
@@ -628,6 +654,8 @@ template BacktrackingMatcher(bool CTregex)
                     case IR.LookbehindEnd:
                     case IR.NeglookbehindEnd:
                     case IR.End:
+                        // cleanup stale stack blocks if any
+                        while (prevStack()) {}
```
Comment (DmitryOlshansky) [inline std/regex/internal/backtracking.d]: I don't get it - CircleCI keeps naging about this line - space after 'while'. But in the latest commit there IS a space between while and paren.

**Comment by DmitryOlshansky** [comment]
Overall - automation is pain in the ass when it doesn't work.

**Comment by JackStouffer** [comment]
Same stable problem will happen here

**Comment by DmitryOlshansky** [comment]
Ping @JackStouffer 
CC @MartinNowak 

This is important bugfix let's not allow it to slip through.

**Comment by quickfur** [comment]
I think it's complaning because it's expecting `{` to begin on a newline after `while(...)`.

**Comment by DmitryOlshansky** [comment]
> I think it's complaning because it's expecting { to begin on a newline after while(...).

Then it has to be fixed not the code. It makes no sense to put {} on the new line just to appease the script.

**Comment by quickfur** [comment]
I agree. But there's also these errors:

````
out/std_regex_package.d(43): Error: undefined identifier 'popFrontN'
out/std_regex_package.d(150): Error: undefined identifier 'Yes.keepSeparators'
out/std_regex_package.d(154): Error: undefined identifier 'Yes.keepSeparators'
````

**Comment by DmitryOlshansky** [comment]
> I agree. But there's also these errors:

These existed on the stable branch and should be fixed separately if at all.

File: std/regex/internal/backtracking.d [hunk 2]
```diff
@@ -1467,6 +1476,8 @@ struct CtContext
             default:
                 assert(0);
             }
+            // cleanup stale stack blocks
+            while(prevStack()){}
```
Comment (quickfur) [inline std/regex/internal/backtracking.d]: You're right, that's the real cause. Sorry for the noise.

Comment (DmitryOlshansky) [inline std/regex/internal/backtracking.d]: Thanks @aG0aep6G, will fix.

**Comment by quickfur** [comment]
You're still missing a space between `while` and the opening `(`.

**Comment by quickfur** [comment]
Btw, what's the feature somebody mentioned where reviewers can edit the PR themselves? That would be very useful here instead of having so many roundtrips back and forth. How does that feature work?

**Comment by DmitryOlshansky** [comment]
> You're still missing a space between while and the opening (.

Aargh! Fixed.

**Comment by quickfur** [comment]
Ah, finally style is passing...

**Comment by wilzbach** [comment]
> Btw, what's the feature somebody mentioned where reviewers can edit the PR themselves? That would be very useful here instead of having so many roundtrips back and forth. How does that feature work?

http://wiki.dlang.org/Guidelines_for_maintainers#write-access-to-prs

**Comment by quickfur** [comment]
But `publictests` is still running into this snag:
````
out/std_regex_package.d(43): Error: undefined identifier 'popFrontN'
out/std_regex_package.d(150): Error: undefined identifier 'Yes.keepSeparators'
out/std_regex_package.d(154): Error: undefined identifier 'Yes.keepSeparators'
````

**Comment by wilzbach** [comment]
> But publictests is still running into this snag:

This test is needed to ensure that all public unittests in dlang.org really compile and don't show import errors to the user.
FWIW the `style` Makefile target is also intended to be run locally - unfortunately only for Posix systems at the moment.

**Comment by quickfur** [comment]
Ah, so the problem is missing imports in a unittest block?

**Comment by wilzbach** [comment]
> Ah, so the problem is missing imports in a unittest block?

Exactly - the script tries to simulate the execution on dlang.org by extracting all public unittests, putting them into a new file (one for each source file, slashes are replaced with underscores) and then subsequently trying to execute the extracted tests.

**Comment by quickfur** [comment]
Hmm, finally figured out how to push commits to somebody else's PR.  That was rather non-obvious!

**Comment by quickfur** [comment]
Woohoo, `publictests` passed!

**Comment by quickfur** [comment]
OK, apparently too much of `std.regex.internal.backtracking` isn't covered by unittests.

Ping @DmitryOlshansky  (btw, I don't know how things are supposed to work after maintainers push commits to the PR; you may have to do a `git pull` to include the unittest fixes I pushed, otherwise circleCI may start complaining again.)

**Comment by wilzbach** [comment]
> Hmm, now it's complaining about code coverage. What's going on?

At the moment I am just at my phone, so a couple of general general points:
- CircleCi runs sth. similar to `make - f posix.mak std/string.test` for all files (i.e. with `-cov`) and then uploads the results To CodeCov 
- the test failures aren't required at the moment (so it's just a warning)
- the browser extension (https://github.com/codecov/browser-extension) saves a couple of clicks as it marks non-covered lines with red in the diff view
- code coverage doesn't with at CTFE

**Comment by DmitryOlshansky** [comment]
> OK, apparently too much of std.regex.internal.backtracking isn't covered by unittests.

I don't buy it. The main reason is I think that coverage is compute by running the tests of the module in isolation. All tests of backtracking are hoisted out to tests.d so any addition of lines will negatively impact coverage in this model.

**Comment by wilzbach** [comment]
> btw, I don't know how things are supposed to work after maintainers push commits to the PR; you may have to do a git pull

I am not sure about the process either, but it's essentially a force-push, so git pulling will set the local branch to the state of the remote branch.

**Comment by wilzbach** [comment]
> I don't buy it. The main reason is I think that coverage is computed by running the tests of the module in isolation. All tests of backtracking are hoisted out to tests.d so any addition of lines will negatively impact coverage in this model.

FYI: The reason for this is that there was some weird, unexplainable behavior of the coverage feature that led to about 2K less covered lines when the test suite was ran with all tests in one go. 
See: https://issues.dlang.org/show_bug.cgi?id=16397

**Comment by quickfur** [comment]
OK, in that case, I'm just gonna push the merge button.  A bugfix for a serious bug as this one shouldn't be held up by these petty things.

**Pull request merged**